src/libstore/gc.cc

```diff
@@ -330,18 +335,184 @@ Roots LocalStore::findRoots()
 }
 
 
+static void readProcLink(int procFd, const char * file, StringSet & paths)
```

Comment:
Because it's a lot simpler than all this passing directory file descriptors around.